econml/tests/test_drlearner.py

```diff
@@ -0,0 +1,558 @@
+from numpy.random import normal, multivariate_normal, binomial
+from sklearn.exceptions import DataConversionWarning
+from sklearn.linear_model import LinearRegression, Lasso, LogisticRegression
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import OneHotEncoder, FunctionTransformer
+from sklearn.model_selection import KFold
+from sklearn.preprocessing import PolynomialFeatures
+from econml.drlearner import DRLearner, LinearDRLearner
+from econml.utilities import shape, hstack, vstack, reshape, cross_product
+from econml.inference import BootstrapInference
+from contextlib import ExitStack
+
+
+# all solutions to underdetermined (or exactly determined) Ax=b are given by A⁺b+(I-A⁺A)w for some arbitrary w
+# note that if Ax=b is overdetermined, this will raise an assertion error
+def rand_sol(A, b):
+    """Generate a random solution to the equation Ax=b."""
+    assert np.linalg.matrix_rank(A) <= len(b)
+    A_plus = np.linalg.pinv(A)
+    x = A_plus @ b
+    return x + (np.eye(x.shape[0]) - A_plus @ A) @ np.random.normal(size=x.shape)
+
+
+class TestDRLearner(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        # Set random seed
+        cls.random_state = np.random.RandomState(12345)
+        # Generate data
+        # DGP constants
+        cls.d = 5
+        cls.n = 1000
+        cls.n_test = 200
+        cls.beta = np.array([0.25, -0.38, 1.41, 0.50, -1.22])
+        cls.heterogeneity_index = 1
+        # Test data
+        cls.X_test = cls.random_state.multivariate_normal(
+            np.zeros(cls.d),
+            np.diag(np.ones(cls.d)),
+            cls.n_test)
+        # Constant treatment effect and propensity
+        cls.const_te_data = TestDRLearner._generate_data(
+            cls.n, cls.d, cls._untreated_outcome,
+            treatment_effect=TestDRLearner._const_te,
+            propensity=lambda x: 0.3)
+        # Heterogeneous treatment and propensity
+        cls.heterogeneous_te_data = TestDRLearner._generate_data(
+            cls.n, cls.d, cls._untreated_outcome,
+            treatment_effect=TestDRLearner._heterogeneous_te,
+            propensity=lambda x: (0.8 if (x[2] > -0.5 and x[2] < 0.5) else 0.2))
+
+    def test_cate_api(self):
+        """Test that we correctly implement the CATE API."""
+        n = 20
+
+        def make_random(is_discrete, d):
+            if d is None:
+                return None
+            sz = (n, d) if d > 0 else (n,)
+            if is_discrete:
+                while True:
+                    arr = np.random.choice(['a', 'b', 'c'], size=sz)
+                    # ensure that we've got at least two of every element
+                    _, counts = np.unique(arr, return_counts=True)
+                    if len(counts) == 3 and counts.min() > 1:
+                        return arr
+            else:
+                return np.random.normal(size=sz)
+
+        d_y = 0
+        is_discrete = True
+        d_t = 1

```

Comment:
Do we not support vectors for T?  I think we should...

Reply:
We do. amended test to test for that